### DISCUSSION TITLE
Cardano Shelley address derivation returning invalid address

### DISCUSSION THREAD

**Discussion by ghost** [trustwallet/wallet-core#2123] [Q&A] [answered]
When using ``wallet.getAddressForCoin()``, and the provided examples in [``CardanoTests.swift``](https://github.com/trustwallet/wallet-core/blob/master/swift/Tests/Blockchains/CardanoTests.swift#L11), an invalid, 115 character address is returned. This is in contrast to the expected 103 character address.

In addition to this, the newly added Cardano signing methods are not defined.

**Comment by hewigovens** [comment]
For some reason Cocoapod release failed for version 2.7.7, could you please try SPM build first? Thanks

**Comment by ghost** [comment]
> For some reason Cocoapod release failed for version 2.7.7, could you please try SPM build first? Thanks

Hmm, I am using the SPM build.

**Comment by hewigovens** [comment]
Try resetting cache and download latest Package.swift or build from release 2.7.7, master points to earlier build

**Comment by murphman300** [comment]
@hewigovens jumping into this convo as we tried to update the package many ways - clearing cache - re-downloading - re-downloading with version pinning in the SPM downloader.

 We're both a bit new with SPM, but we noticed this while pinning the version - with 2.7.7 installed - your Package.swift file has version 2.6.36. We're a bit new to the use of SPM so not sure if this is normal.

Attaching screenshots for context

![Screen Shot 2022-04-06 at 9 40 15 AM](https://user-images.githubusercontent.com/21324779/161988242-dee3beab-0427-469b-8ac8-140469883b66.png)

Also tried this:

![image](https://user-images.githubusercontent.com/21324779/161988994-16b6e808-8c5e-4487-9ba1-a437be16db39.png)



And here are the two versions:

![Screen Shot 2022-04-06 at 9 11 28 AM](https://user-images.githubusercontent.com/21324779/161987903-35212410-e39d-4d2f-9742-de3beb81f72d.png)

**Comment by murphman300** [comment, marked as answer]
We're also not quite sure what you mean by update the Package.swift file, are you suggesting we build your package outside the project and import the build manually?

**Reply by hewigovens** [reply]
Will update master today

**Comment by optout21** [comment]
Could you give an example of an invalid address?

**Reply by ghost** [reply]
addr1ss35adnty3mzqf02vcnmgj4w8w4nt49008gtm6yq29y32v2gvujk3vffanr538z2y0vanxnlqdqm6r3gtwqvs7zgm6vmcugxe3ka5zdne5uu8t

This is a screenshot from Yoroi wallet:

<img width="1279" alt="image" src="https://user-images.githubusercontent.com/99495063/162010591-07d0db6c-f855-495c-b156-a88223076050.png">

**Reply by optout21** [reply]
How did you generate this address?  With wallet-core Cardano code (current master branch) this is NOT a valid address:
 
```
Value of: AddressV3::isValid("addr1ss35adnty3mzqf02vcnmgj4w8w4nt49008gtm6yq29y32v2gvujk3vffanr538z2y0vanxnlqdqm6r3gtwqvs7zgm6vmcugxe3ka5zdne5uu8t")
  Actual: false

Value of: AddressV3::isValidLegacy("addr1ss35adnty3mzqf02vcnmgj4w8w4nt49008gtm6yq29y32v2gvujk3vffanr538z2y0vanxnlqdqm6r3gtwqvs7zgm6vmcugxe3ka5zdne5uu8t")
  Actual: false
```

**Reply by optout21** [reply]
Generated address is shorter (len 103), here is an example:

```
        const auto coin = TWCoinTypeCardano;
        const HDWallet wallet = HDWallet("oil oil oil oil oil oil oil oil oil oil oil oil", "");
        const auto derivationPath = TW::derivationPath(coin);
        PrivateKey privateKey = wallet.getKey(coin, derivationPath);
        std::string address = TW::deriveAddress(coin, privateKey);
        EXPECT_EQ(address, "addr1qy4zgdysk0yt44jksxtzpyrx23dtacvm7lx8as3tcjuypxc82ec75d5j9rayx0pk842rnhgl80z83ad88fpejl405g6sx53rcw");
        EXPECT_TRUE(AddressV3::isValid(address));
```

**Reply by ghost** [reply]
let coinType = try chain.coinType()
       
       guard let wallet = HDWallet(mnemonic: mnemonic, passphrase: "") else {
         throw "error message"
       }
       
       let address = wallet.getAddressForCoin(coin: coinType)

The code above works for all other coins except Cardano.

I also tried using the examples in [CardanoTests.swift](https://github.com/trustwallet/wallet-core/blob/d19d03bab03f47d326b1d284d87753fcfa0509dc/swift/Tests/Blockchains/CardanoTests.swift) and [CoinAddressDerivationTests.swift](https://github.com/trustwallet/wallet-core/blob/d19d03bab03f47d326b1d284d87753fcfa0509dc/swift/Tests/CoinAddressDerivationTests.swift), which both produce the same invalid address.

Is there another way to generate Cardano addresses that I'm unaware of?

**Reply by optout21** [reply]
Please retest, fix has been merged  https://github.com/trustwallet/wallet-core/pull/2142
